compatibility_spark-3/src/main/scala/za/co/absa/hyperdrive/compatibility/impl/DeltaCDCToSnapshotWriter.scala

```diff
@@ -0,0 +1,110 @@
+
+class DeltaCDCToSnapshotWriter(configuration: DeltaCDCToSnapshotWriterConfiguration) extends CompatibleDeltaCDCToSnapshotWriter {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  override def write(dataFrame: DataFrame): StreamingQuery = {
+    dataFrame.writeStream
+      .trigger(configuration.trigger)
+      .outputMode(OutputMode.Append())
+      .option(configuration.checkpointLocationPropName, configuration.checkpointLocation)
+      .options(configuration.extraConfOptions)
+      .foreachBatch((df: DataFrame, batchId: Long) => {
+        if(!DeltaTable.isDeltaTable(df.sparkSession, configuration.destination)) {
+          df.sparkSession
+            .createDataFrame(df.sparkSession.sparkContext.emptyRDD[Row], df.schema)
+            .write
+            .format("delta")
+            .mode(SaveMode.ErrorIfExists)
+            .option("overwriteSchema", "true")
+            .partitionBy(configuration.partitionColumns :_*)
+            .save(configuration.destination)
+        }
+
+        logger.info(s"Writing batchId: $batchId")
+
+        val sortFieldsPrefix = "tmp_hyperdrive_"
+
+        val dataFrameWithSortColumns = getDataFrameWithSortColumns(df, sortFieldsPrefix)
+
+        val originalFieldNames = df.schema.fieldNames.mkString(",")
+        val sortColumnsWithPrefix = configuration.precombineColumns.map(precombineColumn => s"$sortFieldsPrefix$precombineColumn")
+
+        val latestChangeForEachKey = dataFrameWithSortColumns
+          .selectExpr(s"${configuration.keyColumn}", s"struct(${sortColumnsWithPrefix.mkString(",")}, $originalFieldNames) as otherCols" )
+          .groupBy(s"${configuration.keyColumn}")
+          .agg(functions.max("otherCols").as("latest"))
+          .filter(col("latest").isNotNull)
+          .withColumn("latest", new Column(AssertNotNull(col("latest").expr)))
+          .selectExpr( "latest.*")
+          .drop(sortColumnsWithPrefix :_*)
+
+        generateDeltaMerge(latestChangeForEachKey).execute()
+      })
+      .start()
+  }
+
+  private def generateDeltaMerge(latestChanges: DataFrame): DeltaMergeBuilder = {
+    val initialDeltaBuilder = DeltaTable
+      .forPath(configuration.destination)
+      .as("currentTable")
+      .merge(latestChanges.as("changes"), s"currentTable.${configuration.keyColumn} = changes.${configuration.keyColumn}")
+      .whenMatched(s"changes.${configuration.operationColumn} = '${configuration.operationDeleteValue}'")
+      .delete()
+
+    val deltaBuilderWithSortColumns = configuration.precombineColumns.foldLeft(initialDeltaBuilder) { (builder, precombineColumn) =>
+      val order = configuration.precombineColumnsCustomOrder.getOrElse(precombineColumn, Seq.empty[String])
+      order match {
+        case o if o.isEmpty =>
+          builder
+            .whenMatched(s"changes.$precombineColumn > currentTable.$precombineColumn")
+            .updateAll()
+        case o =>
+          val orderString = o.mkString("#")
```

Comment:
In some cases, it will produce wrong results. It is replaced with "#$@". This will work even if "#$@" is part of the precombineColumnsCustomOrder's value. I really doubt there will ever be a value only with "#$@"

Reply:
As discussed fixed